Race condition around payment channel listeners

```
==================
WARNING: DATA RACE
Read at 0x00c00037fcd8 by goroutine 9423:
  github.com/statechannels/go-nitro/node/notifier.(*paymentChannelListeners).Notify()
      /home/runner/work/go-nitro/go-nitro/node/notifier/listeners.go:27 +0x54
  github.com/statechannels/go-nitro/node/notifier.(*ChannelNotifier).NotifyPaymentUpdated()
      /home/runner/work/go-nitro/go-nitro/node/notifier/channel-notifier.go:73 +0x5e4
  github.com/statechannels/go-nitro/node.(*Node).CreateVoucher()
      /home/runner/work/go-nitro/go-nitro/node/node.go:180 +0x337
  github.com/statechannels/go-nitro/rpc.(*RpcServer).registerHandlers.func1.2()
      /home/runner/work/go-nitro/go-nitro/rpc/server.go:128 +0x14e
  github.com/statechannels/go-nitro/rpc.processRequest[go.shape.struct { Amount uint64; Channel github.com/statechannels/go-nitro/types.Destination },go.shape.struct { ChannelId github.com/statechannels/go-nitro/types.Destination; Amount *math/big.Int; Signature github.com/statechannels/go-nitro/crypto.Signature }]()
      /home/runner/work/go-nitro/go-nitro/rpc/server.go:215 +0x528
  github.com/statechannels/go-nitro/rpc.(*RpcServer).registerHandlers.func1()
      /home/runner/work/go-nitro/go-nitro/rpc/server.go:127 +0x4e5
  github.com/statechannels/go-nitro/rpc/transport/http.(*serverHttpTransport).request()
      /home/runner/work/go-nitro/go-nitro/rpc/transport/http/server.go:159 +0x377
  github.com/statechannels/go-nitro/rpc/transport/http.(*serverHttpTransport).request-fm()
      <autogenerated>:1 +0x51
  net/http.HandlerFunc.ServeHTTP()
      /opt/hostedtoolcache/go/1.21.0/x64/src/net/http/server.go:2136 +0x47
  net/http.(*ServeMux).ServeHTTP()
      /opt/hostedtoolcache/go/1.21.0/x64/src/net/http/server.go:2514 +0xbc
  net/http.serverHandler.ServeHTTP()
      /opt/hostedtoolcache/go/1.21.0/x64/src/net/http/server.go:2938 +0x2a1
  net/http.(*conn).serve()
      /opt/hostedtoolcache/go/1.21.0/x64/src/net/http/server.go:2009 +0xc24
  net/http.(*Server).Serve.func3()
      /opt/hostedtoolcache/go/1.21.0/x64/src/net/http/server.go:3086 +0x4f

Previous write at 0x00c00037fcd8 by goroutine 9401:
  github.com/statechannels/go-nitro/node/notifier.(*paymentChannelListeners).Notify()
      /home/runner/work/go-nitro/go-nitro/node/notifier/listeners.go:33 +0x256
  github.com/statechannels/go-nitro/node/notifier.(*ChannelNotifier).NotifyPaymentUpdated()
      /home/runner/work/go-nitro/go-nitro/node/notifier/channel-notifier.go:73 +0x5e4
  github.com/statechannels/go-nitro/node.(*Node).handleEngineEvent()
      /home/runner/work/go-nitro/go-nitro/node/node.go:98 +0x5fd
  github.com/statechannels/go-nitro/node.(*Node).handleEngineEvent-fm()
      <autogenerated>:1 +0x84
  github.com/statechannels/go-nitro/node/engine.(*Engine).run()
      /home/runner/work/go-nitro/go-nitro/node/engine/engine.go:211 +0xa1b
  github.com/statechannels/go-nitro/node/engine.New.func1()
      /home/runner/work/go-nitro/go-nitro/node/engine/engine.go:158 +0x4f

Goroutine 9423 (running) created at:
  net/http.(*Server).Serve()
      /opt/hostedtoolcache/go/1.21.0/x64/src/net/http/server.go:3086 +0x80c
  github.com/statechannels/go-nitro/rpc/transport/http.(*serverHttpTransport).serveHttp()
      /home/runner/work/go-nitro/go-nitro/rpc/transport/http/server.go:92 +0xc4
  github.com/statechannels/go-nitro/rpc/transport/http.NewHttpTransportAsServer.func2()
      /home/runner/work/go-nitro/go-nitro/rpc/transport/http/server.go:85 +0x4f

Goroutine 9401 (running) created at:
  github.com/statechannels/go-nitro/node/engine.New()
      /home/runner/work/go-nitro/go-nitro/node/engine/engine.go:158 +0x81d
  github.com/statechannels/go-nitro/node.New()
      /home/runner/work/go-nitro/go-nitro/node/node.go:57 +0x492
  github.com/statechannels/go-nitro/node_test_test.setupNitroNodeWithRPCClient()
      /home/runner/work/go-nitro/go-nitro/node_test/rpc_test.go:419 +0x[394](https://github.com/statechannels/go-nitro/actions/runs/6424225103/job/17444387168#step:9:395)
  github.com/statechannels/go-nitro/node_test_test.executeNRpcTest()
      /home/runner/work/go-nitro/go-nitro/node_test/rpc_test.go:127 +0x9d2
  github.com/statechannels/go-nitro/node_test_test.executeNRpcTestWrapper.func1()
      /home/runner/work/go-nitro/go-nitro/node_test/rpc_test.go:63 +0x64
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:1648 +0x44
==================
    rpc_test.go:293: Vouchers sent/received
    rpc_test.go:308: Ledger/virtual channels closed
    testing.go:1465: race detected during execution of test
=== NAME  TestRPCWithManualVoucherExchange
    testing.go:1465: race detected during execution of test
--- FAIL: TestRPCWithManualVoucherExchange (1.53s)
    --- FAIL: TestRPCWithManualVoucherExchange/4_clients (1.53s)
=== NAME  
    testing.go:1465: race detected during execution of test
FAIL
FAIL	github.com/statechannels/go-nitro/node_test	13.184s
```

https://github.com/statechannels/go-nitro/actions/runs/6424225103

--- a/node/notifier/listeners.go
+++ b/node/notifier/listeners.go
@@ -48,11 +48,12 @@ func (li *paymentChannelListeners) createNewListener() <-chan query.PaymentChann
 // getOrCreateListener returns the first listener, creating one if none exist.
 func (li *paymentChannelListeners) getOrCreateListener() <-chan query.PaymentChannelInfo {
 	li.listenersLock.Lock()
-	defer li.listenersLock.Unlock()
 	if len(li.listeners) != 0 {
-		return li.listeners[0]
+		l := li.listeners[0]
+		li.listenersLock.Unlock()
+		return l
 	}
-
+	li.listenersLock.Unlock()
 	return li.createNewListener()
 }
 
@@ -110,11 +111,12 @@ func (li *ledgerChannelListeners) createNewListener() <-chan query.LedgerChannel
 // getOrCreateListener returns the first listener, creating one if none exist.
 func (li *ledgerChannelListeners) getOrCreateListener() <-chan query.LedgerChannelInfo {
 	li.listenersLock.Lock()
-	defer li.listenersLock.Unlock()
 	if len(li.listeners) != 0 {
-		return li.listeners[0]
+		l := li.listeners[0]
+		li.listenersLock.Unlock()
+		return l
 	}
-
+	li.listenersLock.Unlock()
 	return li.createNewListener()
 }
 

--- a/node/notifier/listeners.go
+++ b/node/notifier/listeners.go
@@ -12,7 +12,7 @@ type paymentChannelListeners struct {
 	listeners []chan query.PaymentChannelInfo
 	// prev is the previous payment channel info that was sent to the listeners.
 	prev query.PaymentChannelInfo
-	// listenersLock is used to protect against concurrent access to the listeners slice.
+	// listenersLock is used to protect against concurrent access to to sibling struct members.
 	listenersLock *sync.Mutex
 }
 
@@ -24,6 +24,8 @@ func newPaymentChannelListeners() *paymentChannelListeners {
 // Notify notifies all listeners of a payment channel update.
 // It only notifies listeners if the new info is different from the previous info.
 func (li *paymentChannelListeners) Notify(info query.PaymentChannelInfo) {
+	li.listenersLock.Lock()
+	defer li.listenersLock.Unlock()
 	if li.prev.Equal(info) {
 		return
 	}
@@ -45,6 +47,8 @@ func (li *paymentChannelListeners) createNewListener() <-chan query.PaymentChann
 
 // getOrCreateListener returns the first listener, creating one if none exist.
 func (li *paymentChannelListeners) getOrCreateListener() <-chan query.PaymentChannelInfo {
+	li.listenersLock.Lock()
+	defer li.listenersLock.Unlock()
 	if len(li.listeners) != 0 {
 		return li.listeners[0]
 	}
@@ -69,7 +73,7 @@ type ledgerChannelListeners struct {
 	listeners []chan query.LedgerChannelInfo
 	// prev is the previous ledger channel info that was sent to the listeners.
 	prev query.LedgerChannelInfo
-	// listenersLock is used to protect against concurrent access to the listeners slice.
+	// listenersLock is used to protect against concurrent access to sibling struct members.
 	listenersLock sync.Mutex
 }
 
@@ -81,6 +85,8 @@ func newLedgerChannelListeners() *ledgerChannelListeners {
 // Notify notifies all listeners of a ledger channel update.
 // It only notifies listeners if the new info is different from the previous info.
 func (li *ledgerChannelListeners) Notify(info query.LedgerChannelInfo) {
+	li.listenersLock.Lock()
+	defer li.listenersLock.Unlock()
 	if li.prev.Equal(info) {
 		return
 	}
@@ -103,6 +109,8 @@ func (li *ledgerChannelListeners) createNewListener() <-chan query.LedgerChannel
 
 // getOrCreateListener returns the first listener, creating one if none exist.
 func (li *ledgerChannelListeners) getOrCreateListener() <-chan query.LedgerChannelInfo {
+	li.listenersLock.Lock()
+	defer li.listenersLock.Unlock()
 	if len(li.listeners) != 0 {
 		return li.listeners[0]
 	}

